src/api/spec/lib/authenticator_spec.rb

```diff
@@ -95,6 +95,45 @@
                                                                  'but it is not yet approved for the OBS by admin.')
           end
         end
+
+        context 'without kerberos_realm being set' do
+          before do
+            CONFIG.delete 'kerberos_realm'
+            authenticator.extract_user
+          end
+
+          it { expect(CONFIG['kerberos_realm']).to eq('test_realm.com') }
+        end
+
+        context 'without kerberos_service_principal being set' do
+          after do
+            expect { authenticator.extract_user }.to raise_error(Authenticator::AuthenticationRequiredError,
+                                                                 'Kerberos configuration is broken. Principal is empty.')
+          end
+
+          it { CONFIG.delete 'kerberos_service_principal' }
+          it { CONFIG['kerberos_service_principal'] = '' }
```

Comment:
This doesn't make any sense. Please add a test here (expect(foo).to ...) or remove this block

Reply:
> This doesn't make any sense. Please add a test here (expect(foo).to ...) or remove this block

The tests (expect(foo).to ...) are done in the `after` block, placed in line 109. Does it make more sense now?

Reply:
> The tests (expect(foo).to ...) are done in the after block

This also makes no sense to me.. Why not just do it the usual way?
```ruby
before do
  CONFIG.delete 'kerberos_service_principal'
  CONFIG['kerberos_service_principal'] = ''
end

it do
  expect { authenticator.extract_user }.to raise_error(Authenticator::AuthenticationRequiredError,
    'Kerberos configuration is broken. Principal is empty.')
end
```

Reply:
Those are two different tests... it can be together in a `before` block

Reply:
Also, changing the CONF would affect any other test, AFAIK. So better use stub_const:

```
stub_const('CONFIG', 'kerberos_service_principal' => '')
```

Reply:
> Those are two different tests...

Then maybe have two different contexts for it? like
```ruby
context 'deleteing the kerberos_service_principal'
  before do
    CONFIG.delete 'kerberos_service_principal'
  end

  it do
    expect { authenticator.extract_user }.to raise_error(Authenticator::AuthenticationRequiredError,
      'Kerberos configuration is broken. Principal is empty.')
  end
end

context 'setting the kerberos_service_principal to an empty string' do
  before do
    CONFIG['kerberos_service_principal'] = ''
  end

  it do
    expect { authenticator.extract_user }.to raise_error(Authenticator::AuthenticationRequiredError,
      'Kerberos configuration is broken. Principal is empty.')
  end
end
```

Reply:
And why would you delete the principle when you set it again one line after later?;-)

Reply:
In the usual way, you either has to have two context with a before block to set the value of the CONFIG or you have to not have any before block and set the value in each `it` block before the expect... the actual implementation is another way of doing it...

Reply:
This is what the test must achieve:

```
context 'without kerberos_service_principal being set' do
  context 'without kerberos_service_principal key' do
    before { CONFIG.delete 'kerberos_service_principal' }
 
    it do   
      expect { authenticator.extract_user }.to raise_error(Authenticator::AuthenticationRequiredError,
                                                           'Kerberos configuration is broken. Principal is empty.')
    end
  end
 
  context 'with kerberos_service_principal being set to empty string' do
    before { CONFIG['kerberos_service_principal'] = '' }
 
    it do   
      expect { authenticator.extract_user }.to raise_error(Authenticator::AuthenticationRequiredError,
                                                           'Kerberos configuration is broken. Principal is empty.')
    end
  end
end
```

@evanrolfe , your option deletes the key and then assigns it to empty string. We must test both cases. Don't we?

Do you prefer this later option?

Reply:
> And why would you delete the principle when you set it again one line after later?;-)

They are in different `it` groups. `after` blocks would apply for each one. So the test would be run after each `it` code.

Reply:
> bgeuken 24 minutes ago
> And why would you delete the principle when you set it again one line after later?;-)

Nevermind. That was just because Evan used that code in his example

Reply:
> What was proposed with the after block was a refactor of the code, to not repeating the same expect sentence.

I would say the way to not repeat expect statements is with shared_examples. But basically expectations should be inside it blocks, not before/after blocks.

Reply:
what about this?
```ruby
shared_example 'it raises an Authenticator::AuthenticationRequiredError' do
  it do
    expect { authenticator.extract_user }.to raise_error(Authenticator::AuthenticationRequiredError,
      'Kerberos configuration is broken. Principal is empty.')
  end
end

context 'deleteing the kerberos_service_principal'
  before do
    CONFIG.delete 'kerberos_service_principal'
  end

  it_behaves_like 'it raises an Authenticator::AuthenticationRequiredError'
end

context 'setting the kerberos_service_principal to an empty string' do
  before do
    CONFIG['kerberos_service_principal'] = ''
  end

  it_behaves_like 'it raises an Authenticator::AuthenticationRequiredError'
end
 ```

Reply:
I prefer to have the tests in the it blocks. If that means we need two context blocks so be it.

If we want to have this shorter, this could also be done with two it blocks that each contain a stub_const block.

Reasoning behind this is that it's expected to have the actual test in an it block, not in a before or after call.

Reply:
If I didn't understand wrong, we have these options:
* @evanrolfe with [shared_contexts](https://github.com/openSUSE/open-build-service/pull/4211#discussion_r156340256)
* @bgeuken with two it blocks that each contain a stub_const block.

I want an agreement. What are you fine with?

I don't mind to adopt any of them.

Reply:
> Also, changing the CONF would affect any other test, AFAIK. So better use stub_const:
>
> `stub_const('CONFIG', 'kerberos_service_principal' => '')`

`stub_const` is not needed inside this tests, because CONFIG is already stubbed in line 43. So I would prefer not use `stub_const`.

Reply:
> I want an agreement. What are you fine with?

Any way is fine as long as the it follows the rspec format for tests using before, it blocks with expectations, after.